README.md

```diff
@@ -0,0 +1,109 @@
+   * As the `+` is ignored by `cargo`, to re-release a crate with a new version of the API, one would have to increment the patch level. However, that would force all crates to be re-released, even if their API version didn't change at all.
+   * This caused unnecessary spamming of `crates.io`, and the `+` should be a `-` to fix this.
+1. **The 'fields' parameter could not be used to have smaller response types**
+   * Some [Response Types](https://docs.rs/google-sheets4/1.0.10+20190625/google_sheets4/struct.Response.html) are huge, even though with the right `field` setting, one would only receive a fraction of the data. However, one would always have to allocate big structures to with most of the optional fields set to `None`.
+   * 💡Idea 💡: Can [serde(flatten)](https://serde.rs/field-attrs.html#flatten) be used to subdivide the possible field sets in the data structure? Probably it's not known which actual fields belong to each `field` argument.
+
+
+# Technology and Architecture Sketches
+
+Items mentioned below ideally create a link to one of the problems they slove, e.g. `🥵2` , the project goal they support, e.g `🛸3`, or the effective thing they build on (`🌈1`).
+
+## Toolchains
+
+Here is the anticipated tooling. What follows is the list of tools I would add and why.
+
+* **make** - repeats `🌈1` 
+  * I am a fan of simple makefiles, which catch dependencies between files and run a script to generate them. This served _OP_ extremely well.
+  * get parallelization for free, and make transparent which programs to call and how to get work done.
+  * the Makefile serves as hub keeping all commands one would run to interact with the project in any way.
+* **Cargo/Rust** - fixes `🥵4`, support `🌈1`
+  * All work should be done by a Rust binary, which helps keeping things easy with `make`. Previously, the single, magical binary was `python`, and when adding even a single additional Rust tool, one would pay with some complexity. One Rust binary with multiple sub-command seems reasonable.
+* **rust-fmt** - helps remedy `🥵8`
+  * Definitely needed to get idiomatically looking code.
+  * _OP_ didn't have it, it wasn't a real problem, but too much time was spent making things look pretty. With `rust-fmt`, templates can be optimized for maintainability, even if the output doesn't look great initially.
+
+# Development Goals
+
+These should optimize for allowing a pleasant developer experience, at the beginning of the project as well as things stabilize. They should support the project goals or at least not hinder them. For example, settings things up in a way that is hard to use to the average person would be in the way of allowing folks to 'easily' fix issues they encounter.
+
+* **TDD**
+  * Everything done should be driven by at least one test which can be run automatically.
+  * **Byron also thinks that** this is totally doable without breaking into sweat.
```

Comment:
I'll definitely need some guidance here :)

I'm not very good at TDD even in the best circumstances. When it comes to testing something that generates code to interact with an external stateful service that requires credentials I barely even know where to begin.

Reply:
This sort of integration testing is the hard part, and it will never be easy. However, it's possible to nail down everything we know offline with unit-tests, just because our input is json files, and our output is a bunch of transformations on it. These we can test, with snapshot testing for instance. The latter is fantastic if you want to be lazy while still protecting yourself against breakage.

Integration testing to me is less interesting, e.g. testing individual calls to google APIs. It's more relevant to verify user flows. Something I am thinking about is how it's possible to record a flow of operation users would do, and turn these into mocked tests. From time to time the actual responses will be re-recorded, to update the data these mock tests work with. Sometimes there are issues like "I tried to do _this_, but saw the error _that_" - we could reproduce the actions leading to it, fix that, and protect from regression with mock tests, without having to deal with the statefulness of the google apis all the time.

Just thinking out loud, but I think it can be done without dying.